### ISSUE TITLE
Circumventing geo-restrictions

### ISSUE THREAD

**Issue by narcolepticinsomniac** [iv-org/invidious#92]
I thought we were already circumventing geo-blocking, but we're not.

Blocked in the U.S.:
https://invidio.us/watch?v=xpWu4_hp00M

**Comment by omarroth** [comment]
It would be good to start building a list of videos that are "Blocked in US, Blocked in DE, etc".
Do you happen to know in which regions the video linked above is allowed?

**Comment by narcolepticinsomniac** [comment]
Nah, I can't get it to play using any VPN or proxy. Maybe not a great example, as I suppose it's possible that the link is just broken and the geo-restriction aspect isn't really the issue.

**Comment by omarroth** [comment]
Here's an example of a video that is allowed in DE but not US: https://invidio.us/watch?v=rV4Mrl2hQro
It appears that in `/videoplayback` URLs, for example [medium quality for the above video](https://r5---sn-4g5ednsy.googlevideo.com/videoplayback?beids=%5B23745636%5D&key=yt6&ei=RrRoW7KbE8jSgQe517TADA&mv=m&expire=1533610150&c=WEB&initcwndbps=1388750&fvip=6&clen=89361742&ratebypass=yes&gcr=de&signature=7B35049A8F0ED4F5370F00A59B5BCDF4EBC1385E.DF3C2C7A02B3B86685B07198D51AD30BACB0265C&source=youtube&lmt=1529451084101138&gir=yes&ipbits=0&dur=0.000&id=o-AJ5VqTVy7dsdckP3iMqaLVbS_h4zYKmEVkQemRfzSqjz&mime=video%2Fwebm&ms=au%2Conr&requiressl=yes&pl=22&itag=43&mt=1533588498&sparams=clen%2Cdur%2Cei%2Cgcr%2Cgir%2Cid%2Cinitcwndbps%2Cip%2Cipbits%2Citag%2Clmt%2Cmime%2Cmm%2Cmn%2Cms%2Cmv%2Cpl%2Cratebypass%2Crequiressl%2Csource%2Cexpire&mn=sn-4g5ednsy%2Csn-5hne6n7e&mm=31%2C26&ip=136.243.63.59) YouTube now returns 403 when the `ip` parameter does not match the requesting IP.

**Comment by narcolepticinsomniac** [comment]
Haven't I seen something like `bypassip=` params for googlevideo URLs?

**Comment by omarroth** [comment]
You probably saw `ratebypass`, I don't happen to know what it does.
A few quick tests:
- I misspoke, YouTube only cares about `ip=` when the IP is for a video that is not in one of the `allowedRegions`.
- Attempting to access geo-blocked videos through Tor is fine, as long as the exit node is in the correct region.

**Comment by narcolepticinsomniac** [comment]
No, I knew I've seen it. Went through my clipboard manager and found one. IDK if it actually works, but:

```
https://r3---sn-ab5szn7s.googlevideo.com/videoplayback?fvip=3&source=youtube&signature=230E3DC2DF22B86A40021B84ACAD8839AB515422.0D14359F1455503A0BE6096B7B8B1EFB4CD16927&ipbits=0&pl=18&id=o-AIfZC-oGvCCpSym9jRjuh5PzQ3mlPwK2nLKOpIQb6UFj&dur=180.419&mime=video%2Fmp4&key=cms1&ip=2604%3Aa880%3A400%3Ad1%3A%3A180%3Ad001&ei=fGFjW4_5I5SBhwaVtInwDg&c=WEB&requiressl=yes&sparams=dur,ei,expire,id,ip,ipbits,ipbypass,itag,lmt,mime,mip,mm,mn,ms,mv,pl,ratebypass,requiressl,source&itag=22&lmt=1532956145274494&expire=1533261276&ratebypass=yes&redirect_counter=1&rm=sn-ab5eel7s&req_id=ef8047609366a3ee&cms_redirect=yes&ipbypass=yes&mip=74.103.246.134&mm=31&mn=sn-ab5szn7s&ms=au&mt=1533239832&mv=m
```

(towards the end)

**Comment by omarroth** [comment]
Doesn't appear to do anything. A while back I put some work into decrypting/resigning `/videoplayback` URLs but never got anywhere. Notice that in the above URL there is a`key` and `signature`, which prevents tampering.

**Comment by omarroth** [comment]
Working on this right now, currently can get info on the videos linked above. https://invidio.us/watch?v=xpWu4_hp00M appears to be allowed in Germany, Japan, Russia, Thailand (of the regions I've tested). The last thing that needs to be done is to proxy the files themselves, although obviously that will be more bandwidth intensive.

**Comment by dimqua** [comment]
I suggest you to just add this information to the video page, so users can use VPN or proxy to watch videos that are allowed in specific countries.

**Comment by omarroth** [comment]
That's fair enough. I'm already managing downloads for livestreams, and will do the same for #34 and #90, so I thought I might do that here as well. I do plan on adding info to the video page, something like `blacklisted countries` or `whitelisted countries` (depending on which is bigger), and that info is already available though the API.

**Comment by dimqua** [comment]
> regions that make up a significant amount of traffic

How can you measure that?

**Comment by omarroth** [comment]
I think what I mean is the regions that allow a user to watch the largest amount of content. So Germany, US, JP, etc. I can't measure it very well, but I think there is a metric where some percentage of videos are exclusive to region `x`, so it should be added to the list that Invidious tries to bypass.

**Comment by omarroth** [comment]
Attempting to access a geo-blocked file returns a 403 as well, so it appears the only solution is to proxy the file as well. This is also an issue for #90, Google will block access to the file itself in some instances, so this is likely the only solution unfortunately.

**Comment by omarroth** [comment]
Currently the system in place is to cycle through proxies in specific regions, so most of the videos mentioned work (with `dash` enabled so videos are also proxied). The reason it's still open is because there's still a lot of room for improvement.. I expect it to be closed alongside #90. 

The reason geo-blocked videos can take a long time to load is if Invidious can't find a proxy or if none of the regions Invidious tries return a valid response. I would like to fix that before I close this.

**Comment by Perflyst** [comment]
Would it be possible to use Tor for a country from which you dont have a proxy? You can set the exit nodes country if you use tor.

**Comment by omarroth** [comment]
This was suggested in the Matrix server. The issue is that it is not possible to determine where a video is unrestricted without being in an unrestricted region.

The way the current system works is by cycling through proxies in different regions, and finding one where the video is unblocked. Doing this through Tor would put an enormous amount of strain on the network, since it would require building a circuit for every region that is tried.

Also important to note is that video files still have to be proxied from somewhere the video is unblocked. Doing this through Tor is something I'd also rather not do, since it similarly puts strain on the network.

**Comment by elypter** [comment]
you could reduce cycling if you create statistics about how many video requests are blocked in which country

**Comment by jeroen7s** [comment]
This is a really annoying issue, especially with music video's like from Vevo
even trying alternate instances located closer to me don't work.
i'm from Belgium, checking the German, Norwegian, Finnish instance under "alternative instances" doesn't work
the one from the Netherlands seems offline, and there's no Belgian instance

**Comment by jeroen7s** [comment]
proxy-ing does really cause bandwidth constraints though
can't you just catch the video information/url using a proxy, instead of proxy-ing the entire video through invidious servers? that would at least solve the issue that video's don't play, even though they are available in your country when visiting in youtube directly

**Comment by omarroth** [comment]
Sure, it just means you have to track or store the user's region in order to provide correct links. I believe there was also some discussion on this in #460, and I can see it being reasonable to allow a user to "prefer" a different region than where the instance is hosted.

Generally, it's preferable to proxy videos instead of trying to get native links. Mainly because you don't need to manage and retrieve the same video multiple times in different regions which allows you to cut down on the number of requests as well as a couple other optimizations, and since video streams for 1080p (#43) and region-blocked videos need to be proxied through the instance anyway it also makes sense to use the same solution here.

Although I do very much agree that proxying has constraints I believe it is the best solution in this case.

**Comment by tleydxdy** [comment]
perhaps an opt-in "proxy server" config so that my instance can proxy requests for other invidious instance? and pair that with a introduction system so servers can gossip about proxy servers they know, and which region are they.

**Comment by girst** [comment]
@omarroth:  I think I found a marker for geo-locked streams: 

1. get the direct URL to the video (e.g `https://r5---sn-q4fl6nly.googlevideo.com/videoplayback?expire=`...). follow any redirects (redirector.googlevideo.com).
2. parse the query string; if it contains a parameter `gcr` (i got `gcr=ru` in my tests), it is geo/ip locked. 

i only did limited testing; maybe you can try this out on your larger dataset.

**Comment by llacb47** [comment]
In the future when reporting a video you should check which regions it's blocked in with this tool: `http://polsy.org.uk/stuff/ytrestrict.cgi`

For example this video is blocked worldwide except in Brunei: `http://polsy.org.uk/stuff/ytrestrict.cgi?ytid=A-mjDxznYbE&auth=5ba28fe1d53f58118849ffc21bcda842`

**Comment by TheFrenchGhosty** [comment]
A 20$ bounty has been added to this issue.

Anyone opening a PR fixing this issue, will receive 20$ (in BTC) from the Invidious project.

More details: https://github.com/iv-org/invidious/issues/1898

**Comment by unixfox** [comment]
We could use hola proxies which have a lot of available locations: https://github.com/Snawoot/hola-proxy

**Comment by TheFrenchGhosty** [comment]
@unixfox Tor is maybe more reliable than random proxy, maybe?

**Comment by unixfox** [comment]
> @unixfox Tor is maybe more reliable than random proxy, maybe?

You can't choose the country exit node on Tor.

**Comment by outloudvi** [comment]
> You can't choose the country exit node on Tor.

[I think you can](https://tor.stackexchange.com/questions/733/can-i-exit-from-a-specific-country-or-node).

**Comment by syeopite** [comment]
Although Tor is a viable solution, we're going to be consuming a large portion of bandwidth on the already limited Tor network; possibly handicapping connections for those who actually needs Tor. 

Not to mention the fact that Tor is incredibly slow. If this is the solution we decide to use then this should definitely be configurable imo.

**Comment by unixfox** [comment]
> > You can't choose the country exit node on Tor.
> 
> [I think you can](https://tor.stackexchange.com/questions/733/can-i-exit-from-a-specific-country-or-node).

Ok but:
- Tor is slow, and it's a video stream that we are fetching not a simple HTML page. We won't be able to load 720p videos smoothly.
- In how many countries are tor exit nodes hosted? More than 40 countries compared to hola proxy?
- Tor exit nodes are often blocked on Google services.

**Comment by SamantazFox** [comment]
> @unixfox Tor is maybe more reliable than random proxy, maybe?

We shouldn't use Tor. It wasn't designed to handle such a load, and the current network doesn't have the ressources to be used as a video proxy. [See this blog post for more details](https://blog.torproject.org/tors-open-research-topics-2018-edition)

**Comment by unixfox** [comment]
One idea that comes to my mind when discussing with other people on matrix is that we could say to the user that this video is available in this other instance because we fetched the countries from the API of the instances list.

**Comment by syeopite** [comment]
We might have to depend on an external service that categorizes what regions a certain video is allowed in for that considering how YouTube just doesn't return anything sometimes besides a region-locked error message. 

![reason](https://github.com/iv-org/invidious/assets/70992037/953e34b4-e5a8-4fa8-91b3-824b35f8066a)

**Comment by unixfox** [comment]
> We might have to depend on an external service that categorizes what regions a certain video is allowed in for that considering how YouTube sometimes just doesn't return anything sometimes besides a region-locked error message.

Piped have a special external service that use demo official key on YouTube official API for checking the regions where the video is available: https://github.com/TeamPiped/region-restriction-checker

I have also thought of using google, force a country in the settings and search for the video ID in the "videos" tab, then check if the video appears. But it doesn't work well.

And last thoughts, something I have not yet tested, https://www.tunnelbroker.ch/ allows to configure a country when creating a new IPv6 subnet, I wonder if YouTube takes into account this information that is written in the whois or not.

**Comment by syeopite** [comment]
> https://github.com/TeamPiped/region-restriction-checker

This is great! We should find someway of integrating that or something similar we built ourselves into Invidious. I am a bit concerned about calling their official API considering YouTube's recent legal actions but if its an external service and not Invidious directly it should probably be fine.

> https://www.tunnelbroker.ch/  allows to configure a country when creating a new IPv6 subnet...

If that works then that's another alternative and one that is definitely safer than using YouTube's official API.

**Comment by Fijxu** [comment]
@unixfox Do you think this is still relevant? https://github.com/TeamPiped/region-restriction-checker does not work anymore so we can't straight copy what that program did into Invidious, we would have to find our own way. 


> One idea that comes to my mind when discussing with other people on matrix is that we could say to the user that this video is available in this other instance because we fetched the countries from the API of the instances list.

Also, this will require communication between instances and if we want to see if a video is blocked in certain region in another instance, that will require that instance to make a `/player` request (or maybe a `/next` request if geo-restricted information is exposed there). It can also lead to bad actors to use a vulnerable instance with no anti bot (or bypassable anti bot) measures to use one public invidious instance to spam another instance just to check if a video is geo-restricted.

Taking that into account, I don't think this can fixed in the current state :p

(I accidentally clicked **Pin issue**, sorry!)
